### ISSUE TITLE
Allowing initializer words which map to >1 token if num_vectors_per_token supports it

### ISSUE THREAD

**Issue by CodeExplode** [rinongal/textual_inversion#65]
I ran into a situation where the only initializer word I could find which was close to my target was mapped to 2 tokens (using stable diffusion), so have come up with a way which I *think* allows more than 1 token to be the starting basis for textual inversion. This is my first time attempting to write in python or torch, but I think I've got this working (after a very painful afternoon of lessons :P), and just wanted to mention it in case it would be useful to add, or to see if anything obviously wrong about it stands out to the authors.

In embedding_manager.py:

Copied get_clip_token_for_string to get_clip_tokens_for_string, which asserts that (num_vectors_per_token % (nonzerocount-1) == 0), and returns torch.narrow(tokens, 1, 1, count-1) - this should give the tokens for the string, without the trailing 49407's, or the starting 49406, so long as they divide neatly into num_vectors_per_token

Copied get_embedding_for_clip_token to get_embeddings_for_clip_tokens, which returns embedder(tokens)[0]

Changed the code under if initializer_words and idx < len(initializer_words) to:

init_word_tokens = get_clip_tokens_for_string(embedder.tokenizer, initializer_words[idx], num_vectors_per_token)

with torch.no_grad():
  init_word_embeddings = get_embeddings_for_clip_tokens(embedder.transformer.text_model.embeddings, init_word_tokens.cpu() )

num_tokens = torch.count_nonzero(init_word_tokens)
token_params = torch.nn.Parameter( init_word_embeddings.repeat( int(num_vectors_per_token / num_tokens), 1 ), requires_grad=True, )
self.initial_embeddings[placeholder_string] = torch.nn.Parameter(init_word_embeddings.repeat(int(num_vectors_per_token / num_tokens), 1), requires_grad=False, )

Using a num_vectors_per_token of 4 (which maybe 'doubles' my initialization word weight?), my very first training image at 500 steps immediately looks better than anything I've gotten in the first few thousands steps before with the same training data set, and is immediately recognizable as correct, despite the initializing word not being particularly close to the target, only much closer than any words I've found previously. This might be why some targets are able to be mapped quite well, and some have struggled, because it takes an ideal initialization word and many of them would have more than 1 token.

I also spent a long time finding template phrases for personalized.py which generated somewhat correct images when using my initializer word in place of the token, in the hope that it will mostly train the embeddings on what is different.

**Comment by rinongal** [comment]
That's interesting. Did you use the same number of vectors in the runs that took more time to converge? I'd be happy to experiment with this change and add it if it does consistently improve results.

**Comment by CodeExplode** [comment]
Yeah I've tried quite a few times with 4 vectors, since early on I had one get excellent results with 46 training images but it was over-trained on closeup shots of the clothing I'm training it on. Since then I've tried many things from varying dataset sizes (5-120) of different consistencies in quality, mixing in images of all different art style since the leavanny case seemed to benefit from that, different learning rates, etc, but can't seem to get anything decent again.

The new 2-token init word started excellent with 4 or 6 vectors (sometimes producing a training image which looks near perfect straight away), but ended up giving fairly similar results to all my other attempts after a bit of training . The reported loss while training seems to get stuck around 1.1-1.3. However the fact that I was able to get such a good start, which then seemed to go wrong somehow, tells me there's something else I need to look into on top of this. Maybe something in my dataset or other training parameters, or the personalized.py prompts.

**Comment by CodeExplode** [comment]
Another idea, similar to per_image_tokens or progressive_words, is that it could be interesting to train with multiple vectors, and have different data sets and prompts which individual vectors or sub-ranges of the vectors get trained on in isolation, to try to get them to focus on particular elements (such as biases in the data like skin tones). Then some of them might be cut from the final token vectors (e.g. you might train on 8 with the intention to keep 4, and the last 4 are used to try to capture known biases evident in the training data).

Some of the vectors might be used to focus on features like close-up shots, and these sub-vectors could be weighted in the final prompt when using the token (e.g. *[1.0, 1.0, 1.0, 0.5]). So you'd identify parts of your training data set which have that particular feature, and then weight that vector higher when training with them included, and weight it lower when training without them included or referenced as often, to try to get the particular vector to capture that feature.

**Comment by CodeExplode** [comment]
I suppose in the end I'm just reinventing the original tagged images database concept there, and have realized that it would be nice to be able to train multiple embeddings at once which are matched to tags on the images (i.e. they only get used in test generation prompts when an image tag includes the associated term, and maybe are given a user-assigned weighting estimate per image which can be applied to the embedding in the way that prompt weightings have been implemented in some stable diffusion implementations).

Or rather, if we know our data set is going to have some closeup images, distant images, face in frame images, material x images, background colour y images, actor q images, etc, we could tag each training image with those terms and ensure they're included in the test prompt used against that image, and stick to one embedding which doesn't encode all that stuff.
